ocaml/gencert/selfcert.ml

```diff
@@ -74,60 +78,90 @@ let call_openssl args =
   in
   Forkhelpers.execute_command_get_output openssl ~env args
 
-(** [generate_private_key] calls openssl to generate an RSA key of
-  [length] bits. *)
-let generate_private_key length =
+(** [generate_pub_priv_key] calls openssl to generate an RSA key of
+    [length] bits. *)
+let generate_pub_priv_key length =
   let args = ["genrsa"; string_of_int length] in
-  let stdout, _stderr = call_openssl args in
-  stdout
-
-let selfsign issuer extensions key_length days certfile =
-  let rsa =
+  let* rsa_string =
     try
-      generate_private_key key_length
-      |> Cstruct.of_string
-      |> X509.Private_key.decode_pem
-      |> R.failwith_error_msg
-      |> function
-      | `RSA x ->
-          x
+      let stdout, _stderr = call_openssl args in
+      Ok stdout
     with e ->
-      let msg =
-        Printf.sprintf "generating RSA key for %s failed: %s" certfile
-          (Printexc.to_string e)
-      in
-      D.error "%s" msg ; failwith msg
+      let msg = "generating RSA key failed" in
+      D.error "selfcert.ml: %s" msg ;
+      Debug.log_backtrace e (Backtrace.get e) ;
+      R.error_msg msg
+  in
+  let* privkey =
+    rsa_string
+    |> Cstruct.of_string
+    |> X509.Private_key.decode_pem
+    |> R.reword_error (fun _ -> R.msg "decoding private key failed")
+  in
+  let* rsa =
+    try match privkey with `RSA x -> Ok x
+    with _ -> R.error_msg "generated private key does not use RSA"
   in
-  let privkey = `RSA rsa in
   let pubkey = `RSA (Rsa.pub_of_priv rsa) in
+  Ok (privkey, pubkey)
+
+let selfsign' issuer extensions key_length expiration =
+  let* privkey, pubkey = generate_pub_priv_key key_length in
   let req = X509.Signing_request.create issuer privkey in
-  sign days privkey pubkey issuer req extensions >>= fun cert ->
+  let* cert = sign expiration privkey pubkey issuer req extensions in
   let key_pem = X509.Private_key.encode_pem privkey in
   let cert_pem = X509.Certificate.encode_pem cert in
   let pkcs12 =
     String.concat "\n\n" [Cstruct.to_string key_pem; Cstruct.to_string cert_pem]
   in
-  write_certs certfile pkcs12 >>| fun () -> cert
+  Ok (cert, pkcs12)
+
+let selfsign issuer extensions key_length expiration certfile =
+  let* cert, pkcs12 = selfsign' issuer extensions key_length expiration in
+  let* () = write_certs certfile pkcs12 in
+  Ok cert
 
 let host ~name ~dns_names ~ips pemfile =
-  let expire_days = 3650 in
-  let key_length = 2048 in
-  let issuer =
-    [X509.Distinguished_name.(Relative_distinguished_name.singleton (CN name))]
+  let res =
+    let* expiration = expire_in 3650 in
```

Comment:
Using `10 * 365` could make the intent more obvious; likewise a name like `lifetime_days`.